src/builder/QuarkBuilder.sol

```diff
@@ -93,42 +96,80 @@ contract QuarkBuilder {
-                        recipient: transferIntent.recipient,
-                        blockTimestamp: transferIntent.blockTimestamp
-                    })
-                );
-                // TODO: also append a Actions.Action to the actions array.
-                // See: BridgeUSDC TODO for returning a Actions.Action.
+            // Note: Assumes that the asset uses the same # of decimals on each chain
+            uint256 balanceOnDstChain =
+                Accounts.getBalanceOnChain(transferIntent.assetSymbol, transferIntent.chainId, chainAccountsList);
+            uint256 amountLeftToBridge = transferIntent.amount - balanceOnDstChain;
+
+            uint256 bridgeActionCount = 0;
+            // TODO: bridge routing logic (which bridge to prioritize, how many bridges?)
+            // Iterate chainAccountList and find upto 2 chains that can provide enough fund
+            // Backend can provide optimal routes by adjust the order in chainAccountList.
+            for (uint256 i = 0; i < chainAccountsList.length; ++i) {
+                if (amountLeftToBridge == 0) {
+                    break;
+                }
+
+                Accounts.ChainAccounts memory srcChainAccounts = chainAccountsList[i];
+                if (srcChainAccounts.chainId == transferIntent.chainId) {
+                    continue;
+                }
+
+                if (
+                    !BridgeRoutes.canBridge(srcChainAccounts.chainId, transferIntent.chainId, transferIntent.assetSymbol)
+                ) {
+                    continue;
+                }
+
+                Accounts.AssetPositions memory srcAssetPositions =
+                    Accounts.findAssetPositions(transferIntent.assetSymbol, srcChainAccounts.assetPositionsList);
+                Accounts.AccountBalance[] memory srcAccountBalances = srcAssetPositions.accountBalances;
+                // TODO: Make logic smarter. Currently, this uses a greedy algorithm.
+                for (uint256 j = 0; j < srcAccountBalances.length; ++j) {
+                    // TODO: May need to factor in bridge fees for non-CCTP bridges
+                    if (srcAccountBalances[j].balance < BRIDGE_MINIMUM_AMOUNT_USDC) {
+                        continue;
+                    }
+
+                    if (bridgeActionCount >= MAX_BRIDGE_ACTION) {
+                        revert TooManyBridgeOperations();
+                    }
+
+                    uint256 amountToBridge = srcAccountBalances[j].balance >= amountLeftToBridge
+                        ? amountLeftToBridge
+                        : srcAccountBalances[j].balance;
+                    amountLeftToBridge -= amountToBridge;
+
+                    if (payment.isToken) {
+                        // TODO: wrap around paycall
+                    } else {
+                        // TODO: Should change to handle non-USDC bridges later on
+                        (quarkOperations[actionIndex], actions[actionIndex]) = Actions.bridgeUSDC(
```

Comment:
Is this actually `bridgeUSDC` or is this `bridgeCctp`? E.g. if we expanded to two USDC bridges, would this function be accurately named?

Reply:
Done. I generalized this from `bridgeUSDC` to `bridgeAsset`. The logic for `Actions.bridgeAsset` now can handle any type of asset. For unsupported assets (e.g. non-USDC as of now), the function will revert.